File: src/commands/update.ts
import { TomlParser } from '../util/toml.js'
import { RegistryClient } from '../registry/client.js'
import { FileUtils } from '../util/fs.js'
import { Lockfile, LockfileEntry } from '../lockfile.js'
import { Semver } from '../model/semver.js'
import { SemverRange } from '../model/semver.js'
import { Spinner } from '../ui/spinner.js'
import { cli } from '../ui/colors.js'
import path from 'path'
import fs from 'fs'

export interface UpdateOptions {
  dryRun?: boolean
  verbose?: boolean
}

export class UpdateCommand {
  constructor(private projectDir: string) {}

  async run(pkgNames: string[], opts: UpdateOptions = {}): Promise<void> {
    const inkPackageTomlPath = path.join(this.projectDir, 'ink-package.toml')
    if (!fs.existsSync(inkPackageTomlPath)) {
      console.error('No ink-package.toml found. Run `quill init` or `quill new` first.')
      process.exit(1)
    }

    const manifest = TomlParser.read(inkPackageTomlPath)
    const deps: Record<string, string> = manifest.dependencies ?? {}

    if (Object.keys(deps).length === 0) {
      console.log('No dependencies to update.')
      return
    }

    // Filter to requested packages, or all if none specified
    const targets = pkgNames.length > 0
      ? pkgNames.filter(n => {
          if (!deps[n]) { console.warn(`${n} is not in dependencies, skipping.`); return false }
          return true
        })
      : Object.keys(deps)

    if (targets.length === 0) return

    const client = new RegistryClient()
    const spinner = new Spinner()

    spinner.start('Fetching registry index...')
    const index = await client.fetchIndex()
    spinner.succeed('Registry index fetched.')

    const packagesDir = path.join(this.projectDir, 'packages')
    const cacheDir = path.join(this.projectDir, '.quill-cache')
    const lockedPkgs: Record<string, LockfileEntry> = {}
    const updatedDeps = { ...deps }
    let updatedCount = 0

    for (const depName of targets) {
      const currentRange = deps[depName]
      const pkgVersion = client.findBestMatch(index, depName, currentRange)
      if (!pkgVersion) {
        console.error(`No version of ${depName} satisfies ${currentRange}`)
        continue
      }

      const pkgDir = path.join(packagesDir, depName.replace('/', '-'))

      let installedVersion: string | null = null
      const installedManifest = path.join(pkgDir, 'ink-package.toml')
      if (fs.existsSync(pkgDir) && fs.existsSync(installedManifest)) {
        installedVersion = TomlParser.read(installedManifest).version ?? null
      }

      if (installedVersion === pkgVersion.version) {
        console.log(`${cli.muted('─')} ${depName} v${pkgVersion.version} is already up to date.`)
        updatedDeps[depName] = deps[depName]
      } else {
        if (opts.dryRun) {
          console.log(`[dry-run] Would update ${depName} v${installedVersion ?? '?'} → v${pkgVersion.version}`)
          if (opts.verbose) {
            console.log(`  URL: ${pkgVersion.url}`)
          }
          updatedDeps[depName] = deps[depName]
        } else {
          if (installedVersion) fs.rmSync(pkgDir, { recursive: true, force: true })
          FileUtils.ensureDir(cacheDir)
          const tarball = path.join(cacheDir, `${depName.replace('/', '-')}-${pkgVersion.version}.tar.gz`)
          spinner.start(`Updating ${depName} → v${pkgVersion.version}...`)
          if (opts.verbose) console.log(`  URL: ${pkgVersion.url}`)
          await FileUtils.downloadFile(pkgVersion.url, tarball)
          await FileUtils.extractTarGz(tarball, pkgDir)
          spinner.succeed(`${depName} updated to v${pkgVersion.version}.`)
          updatedCount++
          updatedDeps[depName] = `^${pkgVersion.version}`
        }
      }

      lockedPkgs[`${depName}@${pkgVersion.version}`] = new LockfileEntry(pkgVersion.version, pkgVersion.url)
    }

    // Carry over untouched deps
    for (const [dep, range] of Object.entries(deps)) {
      if (!targets.includes(dep)) {
        const pv = client.findBestMatch(index, dep, range)
        if (pv) lockedPkgs[`${dep}@${pv.version}`] = new LockfileEntry(pv.version, pv.url)
      }
    }

    if (opts.dryRun) {
      const lockfile = new Lockfile(client.registryUrl, lockedPkgs)
      lockfile.write(path.join(this.projectDir, 'quill.lock'))
      console.log(`\n${cli.muted('[ dry-run: lockfile updated but no packages modified ]')}`)
      return
    }

    if (updatedCount > 0) {
      const updated = { ...manifest, dependencies: updatedDeps }
      fs.writeFileSync(inkPackageTomlPath, TomlParser.write(updated))
    }

    const lockfile = new Lockfile(client.registryUrl, lockedPkgs)
    lockfile.write(path.join(this.projectDir, 'quill.lock'))

    if (updatedCount === 0) {
      console.log('All packages are up to date.')
    } else {
      console.log(`Updated ${updatedCount} package(s).`)
    }
  }
}

File: src/commands/update.rs
use async_trait::async_trait;
use std::collections::BTreeMap;

use crate::commands::Command;
use crate::context::Context;
use crate::error::Result;
use crate::manifest::Lockfile;
use crate::registry::RegistryClient;
use crate::resolve;

pub struct Update {
    pub precision: Option<String>,
    pub recursive: bool,
    pub packages: Vec<String>,
}

#[async_trait]
impl Command for Update {
    async fn execute(&self, ctx: &Context) -> Result<()> {
        let manifest = ctx.manifest.as_ref().ok_or_else(|| {
            crate::error::QuillError::ManifestNotFound {
                path: ctx.project_dir.join("ink-manifest.toml"),
            }
        })?;

        let lockfile = ctx.lockfile.as_ref().ok_or_else(|| {
            crate::error::QuillError::io_error(
                "lockfile not found",
                std::io::Error::new(std::io::ErrorKind::NotFound, "lockfile not found")
            )
        })?;

        let registry_url = &ctx.registry_url;
        let client = RegistryClient::new(registry_url);

        // Fetch latest registry index
        let index = client.fetch_index().await?;

        // Determine which packages to update
        let packages_to_update: Vec<String> = if self.packages.is_empty() {
            // Update all dependencies
            manifest.dependencies.keys().cloned().collect()
        } else {
            self.packages.clone()
        };

        // Build updated dependencies map with new version ranges
        let mut updated_deps: BTreeMap<String, String> = manifest.dependencies.clone();

        for pkg_name in &packages_to_update {
            if let Some(current_range) = manifest.dependencies.get(pkg_name) {
                // Find latest version matching the current range precision
                if let Some((version_str, _pkg)) = index.find_best_match(pkg_name, current_range) {
                    updated_deps.insert(pkg_name.clone(), format!("^{}", version_str));
                    println!("Updated {}: {} -> ^{}", pkg_name, current_range, version_str);
                }
            }
        }

        // Re-resolve with updated dependencies
        let resolved = resolve::resolve_transitive(&index, &updated_deps)?;

        // Update lockfile
        let mut packages: BTreeMap<String, crate::manifest::LockedPackage> = BTreeMap::new();
        for (name, resolved_pkg) in &resolved {
            packages.insert(
                name.clone(),
                crate::manifest::LockedPackage {
                    version: resolved_pkg.version.clone(),
                    resolution_source: registry_url.clone(),
                    dependencies: resolved_pkg.dep_keys.clone(),
                },
            );
        }

        let updated_lockfile = Lockfile {
            version: lockfile.version,
            registry: registry_url.clone(),
            packages,
        };

        // Save updated lockfile
        let lockfile_path = ctx.project_dir.join("quill.lock");
        updated_lockfile.save(&lockfile_path)?;

        // Save updated manifest
        let mut updated_manifest = manifest.clone();
        updated_manifest.dependencies = updated_deps;

        let manifest_path = ctx.project_dir.join("ink-manifest.toml");
        let content = toml::to_string_pretty(&updated_manifest)
            .map_err(|e| crate::error::QuillError::RegistryAuth {
                message: format!("failed to serialize manifest: {}", e),
            })?;
        std::fs::write(&manifest_path, content)
            .map_err(|e| crate::error::QuillError::io_error("failed to write manifest", e))?;

        println!("Updated dependencies");
        Ok(())
    }
}
